### PR TITLE
✅ Implementation of unit tests

### LINKED ISSUE #1: 🧪 Unit Tests Implementation
# 🧪 Implement unit tests for core functions

## 🎯 Objective

Implement comprehensive unit tests for the L1 distance calculation functions and image preparation utilities.

### L1 Distance Functions
- [ ] Implement tests for `l1_generic` function
- [ ] Implement tests for `l1_x86_sse2` function (x86_64 architecture)
- [ ] Implement tests for `l1_neon` function (ARM architecture)
- [ ] Ensure tests validate correctness of distance calculations
- [ ] Test edge cases (empty arrays, single pixel, large arrays)

### Image Preparation Functions
- [ ] Implement tests for `prepare_target` function
  - [ ] Verify correct image resizing
  - [ ] Validate output dimensions
  - [ ] Test different input image sizes
- [ ] Implement tests for `prepare_tiles` function
  - [ ] Verify correct tile sizing
  - [ ] Validate tile dimensions
  - [ ] Test with different tile configurations

## ✅ Acceptance Criteria

- All unit tests pass with `cargo test`
- Tests cover main functionality and edge cases
- Code coverage for tested functions > 80%
- Tests are well-documented with clear assertions
- Tests run successfully on both x86_64 and aarch64 (if applicable)

## 💡 Implementation Hints

- Use Rust's `#[test]` attribute
- Consider using `#[cfg(target_arch = "...")]` for architecture-specific tests
- Use assertion macros like `assert_eq!`, `assert!`
- Group related tests in modules using `mod tests`

### LINKED ISSUE #2: 🔗 Integration Tests Implementation
# 🔗 Implement integration tests with ground truth validation

## 🎯 Objective

Implement integration tests in `tests/temp.rs` to validate the complete mosaic generation pipeline using ground truth images.

## 📋 Tasks

- [ ] Set up integration test file `tests/temp.rs`
- [ ] Implement test for `compute_mosaic` function
- [ ] Test mosaic generation for **generic** implementation
- [ ] Test mosaic generation for **x86_64 SSE2** implementation
- [ ] Test mosaic generation for **ARM NEON** implementation
- [ ] Compare generated mosaic with provided ground truth image
- [ ] Implement pixel-by-pixel comparison or similarity metric
- [ ] Add test assertions for acceptable differences (tolerance)
- [ ] Document test setup and expected outcomes

## ✅ Acceptance Criteria

- Integration tests run successfully with `cargo test`
- Tests validate output against ground truth image from `assets/` directory
- All architecture-specific implementations are tested (where applicable)
- Tests include proper error handling
- Clear documentation of acceptable tolerance thresholds
- Tests pass on both x86_64 and aarch64 architectures

## 💡 Implementation Hints

- Place integration tests in `tests/` directory (separate from unit tests)
- Use `image` crate to load and compare images
- Consider using `#[cfg(target_arch = "...")]` for architecture-specific tests
- Use the provided assets (target image, tiles, ground truth) for testing
- Implement a helper function to calculate image similarity
- Acceptable tolerance might be needed due to floating-point arithmetic

### PR DESCRIPTION
**Pull request by Mouadfrh**
Rajouter des tests unitaires pour tester le fonctionnement des fonctions l1_neon, l1_x86_sse2, l1_generic, prepare_target et prepare_tiles.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -434,28 +434,120 @@ pub fn compute_mosaic(args: Options) {
 }
 
 fn main() {
-    let args = Options::parse();
+    let args = Options::parse();
```
Comment (Sachatms) [inline src/main.rs:437]: no change necessary here
```suggestion
    let args = Options::parse();
```

File: src/main.rs [hunk 2]
```diff
@@ -434,28 +434,120 @@ pub fn compute_mosaic(args: Options) {
...
+
+        // Expected distance = sum of absolute differences for each pixel and each channel
+        // Each pixel contributes (5 + 5 + 5) = 15 → 4x4 pixels → 16 * 15 = 240
+        let expected_distance = 4 * 4 * (5 + 5 + 5);
+        (im1, im2, expected_distance)
+    }
+
+    // ---- L1 TESTS ----
+    // These tests verify that L1 distance computation gives the expected result
+    // across different CPU architectures and instruction sets (AVX2, SSE2, NEON).
```
Comment (Sachatms) [inline src/main.rs]: avoid useless comment
```suggestion
    
```

File: src/main.rs [hunk 3]
```diff
@@ -434,28 +434,120 @@ pub fn compute_mosaic(args: Options) {
...
+
+        // Fill first image with RGB (10, 20, 30)
+        for pixel in im1.pixels_mut() {
+            *pixel = image::Rgb([10, 20, 30]);
+        }
+
+        // Fill second image with RGB (15, 25, 35)
+        for pixel in im2.pixels_mut() {
+            *pixel = image::Rgb([15, 25, 35]);
+        }
```
Comment (Sachatms) [inline src/main.rs]: It's not possible to use an iterator instead?

File: src/main.rs [hunk 4]
```diff
@@ -434,28 +434,120 @@ pub fn compute_mosaic(args: Options) {
...
+                // Test optimized AVX2 version
+                assert_eq!(l1_x86_avx2(&im1, &im2), expected);
+            } else if is_x86_feature_detected!("sse2") {
+                // Test SSE2 fallback version
+                assert_eq!(l1_x86_sse2(&im1, &im2), expected);
+            } else {
+                // If no SIMD instruction set is available, skip test
+                assert!(true);
+            }
+        }
```
Comment (Sachatms) [inline src/main.rs:471]: why unsafe?

File: src/main.rs [hunk 5]
```diff
@@ -434,28 +434,120 @@ pub fn compute_mosaic(args: Options) {
...
 
     #[test]
     #[cfg(target_arch = "aarch64")]
     fn unit_test_aarch64() {
-        assert!(true);
+        let (im1, im2, expected) = create_test_images();
+        unsafe {
+            // Test NEON implementation for ARM CPUs
+            assert_eq!(l1_neon(&im1, &im2), expected);
+        }
```
Comment (Sachatms) [inline src/main.rs:480]: same here

File: src/main.rs [hunk 6]
```diff
@@ -434,28 +434,120 @@ pub fn compute_mosaic(args: Options) {
...
     fn unit_test_generic() {
-        // TODO
-        assert!(true);
+        let (im1, im2, expected) = create_test_images();
+        // Test generic (non-optimized) version
+        assert_eq!(l1_generic(&im1, &im2), expected);
+    }
+
+    // ---- TILE TEST ----
+    // This test checks that the tile preparation function loads and resizes images correctly.
```
Comment (Sachatms) [inline src/main.rs]:
```suggestion
```

File: src/main.rs [hunk 7]
```diff
@@ -434,28 +434,120 @@ pub fn compute_mosaic(args: Options) {
...
+            assert_eq!(tile.width(), 4);
+            assert_eq!(tile.height(), 4);
+        }
+
+        fs::remove_dir_all(dir).unwrap();
+    }
+
+    // ---- TARGET TEST ----
+    // This test checks that the target image is properly resized and prepared.
+
```
Comment (Sachatms) [inline src/main.rs]:
```suggestion
```

File: src/main.rs [hunk 8]
```diff
@@ -434,28 +434,120 @@ pub fn compute_mosaic(args: Options) {
...
+        let tile_size = Size { width: 4, height: 4 };
+        let result = prepare_target(path, 1, &tile_size).unwrap();
+
+        // The target should be correctly resized to 4x4
+        assert_eq!(result.width(), 8);
+        assert_eq!(result.height(), 8);
+
+        fs::remove_file(path).unwrap();
     }
```
Comment (Sachatms) [inline src/main.rs:532]: prepare_target, prepare_tile, give a Result<> as output and you've only tested the Ok() output, add more test to test the error behaviour too?

**Review by Sachatms — CHANGES_REQUESTED** [review]
try to cover every possibility of output from every function and have a look on the possible fail or not in function of parameter
